### PR TITLE
many: add transactional flag to snapd API

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Add transactional flag to snapd API. This is a continuation of https://github.com/snapcore/snapd/pull/11271, and it is the second of a series of three (see https://github.com/snapcore/snapd/pull/11255 for the original PR).

### CODE REVIEW

File: cmd/snap/cmd_snap_op.go [hunk 1]
```diff
@@ -826,17 +828,18 @@ func (x *cmdRefresh) Execute([]string) error {
...
+		IgnoreValidation: x.IgnoreValidation,
+		IgnoreRunning:    x.IgnoreRunning,
+		Revision:         x.Revision,
+		CohortKey:        x.Cohort,
+		LeaveCohort:      x.LeaveCohort,
+		Transactional:    x.Transactional,
+	}
+	x.setModes(opts)
 	if len(names) == 1 {
-		opts := &client.SnapOptions{
```
Comment (bboozzoo) [inline cmd/snap/cmd_snap_op.go:830]: I wonder if it wouldn't be clearer if we had:
```
if len(names) == 1 {
	opts := &client.SnapOptions{
		Amend:            x.Amend,
		Channel:          x.Channel,
		IgnoreValidation: x.IgnoreValidation,
		IgnoreRunning:    x.IgnoreRunning,
		Revision:         x.Revision,
		CohortKey:        x.Cohort,
		LeaveCohort:      x.LeaveCohort,
		Transactional:    x.Transactional,
	}
	x.setModes(opts)
	return x.refreshOne(names[0], opts)
}
// only transactional flag has meaning when refreshing
// with many snaps
opts := &client.SnapOptions{
	Transactional:    x.Transactional,
}
...
```

File: client/snap_op.go [hunk 2]
```diff
@@ -197,8 +204,13 @@ func (client *Client) doMultiSnapActionFull(actionName string, snaps []string, o
 		Snaps:  snaps,
 	}
 	if options != nil {
+		if options.Dangerous {
+			return nil, "", ErrDangerousNotApplicable
```
Comment (bboozzoo) [inline client/snap_op.go:208]: This introduces a change in the behavior of snap install, previously `snap install foo bar --dangerous` would just silently ignore the dangerous flag what means that people could have scripted such operations and they did not fail. Now we actively return `error: dangerous option only meaningful when installing from a local file`. Perhaps we should continue ignoring most flags here, and tweak the failing unit test that the flag (and maybe some other flags) is indeed ignored.

File: daemon/api_snaps_test.go [hunk 3]
```diff
@@ -626,6 +626,45 @@ func (s *snapsSuite) TestRefreshAllRestoresValidationSets(c *check.C) {
...
+	c.Check(res.Summary, check.Equals, `Refresh snaps "foo", "bar"`)
+	c.Check(res.Affected, check.DeepEquals, inst.Snaps)
+	c.Check(refreshSnapAssertions, check.Equals, true)
+	c.Assert(refreshAssertionsOpts, check.NotNil)
+	c.Check(refreshAssertionsOpts.IsRefreshOfAllSnaps, check.Equals, false)
+
+	c.Check(calledFlags.Transactional, check.Equals, true)
+}
+
+func (s *snapsSuite) TestRefreshManyTransactionally(c *check.C) {
```
Comment (pedronis) [inline daemon/api_snaps_test.go:667]: it seems that this name should apply to the new test?

File: daemon/api_snaps_test.go [hunk 4]
```diff
@@ -626,6 +626,45 @@ func (s *snapsSuite) TestRefreshAllRestoresValidationSets(c *check.C) {
 }
 
 func (s *snapsSuite) TestRefreshMany(c *check.C) {
```
Comment (pedronis) [inline daemon/api_snaps_test.go:628]: this should be renamed instead

File: cmd/snap/cmd_snap_op.go (hunk 1 above)
Comment (pedronis) [reply cmd/snap/cmd_snap_op.go:830]: I think I agree with this suggestion

File: client/snap_op.go [hunk 5]
```diff
@@ -55,30 +56,38 @@ func writeFieldBool(mw *multipart.Writer, key string, val bool) error {
 	return mw.WriteField(key, "true")
 }
 
+type field struct {
+	field string
+	value bool
+}
+
+func writeFields(mw *multipart.Writer, fields *[]field) error {
```
Comment (pedronis) [inline client/snap_op.go:64]: pointer to slices are very unusual, unless you need to write over fields itself, I would expect []field here

File: client/snap_op.go (hunk 2 above)
Comment (pedronis) [reply client/snap_op.go:208]: we might want to consider this change, probably separately from the rest of the feature though. It's unfortunate that we have simply ignored all flags so far for the many case.

**Review by pedronis — COMMENTED** [review]
did a first pass

File: client/snap_op.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply client/snap_op.go]: I've changed it so it does not error out now and added a TODO

File: daemon/api_snaps_test.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply daemon/api_snaps_test.go]: Oops, yeah, thanks for pointing out

**Comment by alfonsosanchezbeato** [comment]
Thanks for the comments, I've pushed now changes addressing them.

**Review by bboozzoo — APPROVED** [review]
Thanks

File: client/snap_op.go [hunk 6]
```diff
@@ -204,9 +204,7 @@ func (client *Client) doMultiSnapActionFull(actionName string, snaps []string, o
 		Snaps:  snaps,
 	}
 	if options != nil {
-		if options.Dangerous {
-			return nil, "", ErrDangerousNotApplicable
-		}
+		// TODO consider returning error when options.Dangerous is set
```
Comment (pedronis) [inline client/snap_op.go:207]: thanks for adding this

nitpick, we tend to format these as: // TODO: ...

with a colon (:)

**Review by pedronis — APPROVED** [review]
thank you, TODO formatting nitpick

**Pull request merged**